libs/jarstore.js

```diff
@@ -0,0 +1,196 @@
+    request.onupgradeneeded = function(event) {
+      var database = request.result;
+      var transaction = request.transaction;
+
+      var version = event.oldVersion;
+      (function next() {
+        if (version < event.newVersion) {
+          upgrade[version + "to" + ++version](database, transaction, next);
+        }
+      })();
+    };
+
+    request.onsuccess = function() {
+      database = request.result;
+      resolve();
+    };
+  });
+
+  function addBuiltIn(jarName, jarData) {
+    var zip = new ZipFile(jarData, false);
+
+    jars.set(jarName, {
+      directory: zip.directory,
+      isBuiltIn: true,
+    });
+  }
+
+  function installJAR(jarName, jarData, jadData) {
+    return openDatabase.then(function() {
+      return new Promise(function(resolve, reject) {
+        var zip = new ZipFile(jarData, true);
+
+        var transaction = database.transaction(OBJECT_STORE, "readwrite");
+        var objectStore = transaction.objectStore(OBJECT_STORE);
+        var request = objectStore.put({
+          jar: zip.directory,
+          jad: jadData ? jadData : null,
+        }, jarName);
+
+        request.onerror = function() {
+          console.error("Error installing: " + request.error.name);
+          reject(request.error.name);
+        };
+
+        transaction.oncomplete = function() {
+          jars.set(jarName, {
+            directory: zip.directory,
+            isBuiltIn: false,
+          });
+          resolve();
+        };
+      });
+    });
+  }
+
+  function loadJAR(jarName) {
+    return openDatabase.then(function() {
+      return new Promise(function(resolve, reject) {
+        var transaction = database.transaction(OBJECT_STORE, "readonly");
+        var objectStore = transaction.objectStore(OBJECT_STORE);
+        var request = objectStore.get(jarName);
+
+        request.onerror = function() {
+          console.error("Error loading: " + request.error.name);
+          reject(request.error.name);
+        };
+
+        transaction.oncomplete = function() {
+          if (request.result) {
+            jars.set(jarName, {
+              directory: request.result.jar,
+              isBuiltIn: false,
+            });
+
+            if (request.result.jad) {
+              jad = request.result.jad;
+            }
+
+            resolve(true);
+          } else {
+            resolve(false);
+          }
+        };
+      });
+    });
+  }
+
+  function loadFileFromJAR(jarName, fileName) {
+    var jar = jars.get(jarName);
+    if (!jar) {
+      return null;
+    }
+
+    var entry = jar.directory[fileName];
+
+    if (!entry) {
+      return null;
+    }
+
+    var bytes;
+    if (entry.compression_method === 0) {
+      bytes = entry.compressed_data;
+    } else if (entry.compression_method === 8) {
+      bytes = inflate(entry.compressed_data);
+    } else {
+      return null;
+    }
+
+    if (jar.isBuiltIn) {
+      return bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength)
```

Comment:
This is pre-existing code, I can't return bytes.buffer because the code that uses this function is expecting an ArrayBuffer of a single file (while bytes.buffer for built-in JARs is the ArrayBuffer of the entire JAR)